### PR TITLE
Security patch for using remote data as params

### PR DESCRIPTION
**Pull request by eugenepeh**
```
Many of our operations use terminal utilities.

This makes our application prone to security issues when we use remote
data as parameters in the terminal.

As a step towards user's safety, let's validate the data to be free of
malicious input before using them in our operations.
```

### CODE REVIEW

File: src/main/java/reposense/model/Author.java [hunk 1]
```diff
@@ -8,21 +8,91 @@
...
+     * Returns true if a given string is a valid name.
+     */
+    private static boolean isValidName(String name) {
+        return name.matches(NAME_VALIDATION_REGEX);
+    }
+
+    /**
+     * Returns true if all the strings in the {@code ignoreGlobList} only contains commonly used glob patterns.
+     */
+    private static boolean validateIgnoreGlobs(List<String> ignoreGlobList) {
```
Comment (yamidark) [inline src/main/java/reposense/model/Author.java]: no need to be a `boolean`?
Instead, should mention that it throws `IllegalArgumentException`?

File: src/main/java/reposense/model/Author.java [hunk 2]
```diff
@@ -8,21 +8,91 @@
...
+                throw new IllegalArgumentException(String.format(MESSAGE_UNCOMMON_GLOB_PATTERN, glob));
+            }
+        }
+        return true;
+    }
+
+    /**
+     * Returns true if all the strings in the {@code authorAliases} are valid names.
+     */
+    private static boolean validateAuthorAliases(List<String> authorAliases) {
```
Comment (yamidark) [inline src/main/java/reposense/model/Author.java]: Same here

File: src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+			"githubId": "lithiumlkid",
+			"displayName": "Ahm",
+			"authorNames": ["Ahmad Syafiq"],
+			"ignoreGlobList": ["*.aa1", "**.aa2", "**.java"]
+		},
+		{
+			"githubId": "codeeong",
+			"displayName": "Cod",
+			"authorNames": ["Codee; ls"],
+			"ignoreGlobList": ["**[!(.md)]"]
```
Comment (yamidark) [inline src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json]: can each `_configjson` test file contain just that 1 author that has the invalid character, so its easier to tell which is invalid?

File: src/test/java/reposense/model/StandaloneConfigTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+            .getResource("StandaloneConfigTest/invalidIgnoreGlob_config.json").getFile()).toPath();
+
+    @Test
+    public void standaloneConfig_validJson_success() throws IOException {
+        StandaloneConfig standaloneConfig = new StandaloneConfigJsonParser().parse(VALID_CONFIG);
+        config.update(standaloneConfig);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void standaloneConfig_invalidGitIdJson_throwIllegalArgumentException() throws IOException {
```
Comment (yamidark) [inline src/test/java/reposense/model/StandaloneConfigTest.java]: Since we are already adding tests for `StandaloneConfig` already, can we add additional tests, like non-json file, contain invalid field names?

File: src/test/java/reposense/model/AuthorTest.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author("lithiumlkid & echo hi"));
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author("lithiumlkid ; echo hi"));
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author("lithiumlkid | rm -rf /"));
+    }
+
+    @Test
+    public void constructor_validGitHubId_success() {
+        String gitId = "lithiumlkid";
+        Author author = new Author(gitId);
+        assertEquals(author.getGitId(), gitId);
```
Comment (yamidark) [inline src/test/java/reposense/model/AuthorTest.java]: `Assert.assertEquals`?
Also, `gitId` is expected so should be LHS?

File: src/test/java/reposense/model/AuthorTest.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+        gitId = "LAPTOP-7KFM2KSP\\User";
+        author = new Author(gitId);
+        assertEquals(author.getGitId(), gitId);
```
Comment (yamidark) [inline src/test/java/reposense/model/AuthorTest.java]: Same as above

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 7]
```diff
@@ -126,24 +126,20 @@ public static void setFormatsToRepoConfigs(List<RepoConfiguration> configs, List
...
-
-            this.setAuthorDisplayName(author, displayName);
-            this.addAuthorAliases(author, Arrays.asList(sa.getGithubId()));
-            this.addAuthorAliases(author, sa.getAuthorNames());
+            Author author = new Author(sa);
+            author.appendIgnoreGlobList(ignoreGlobList);
+
+            this.authorList.add(author);
+
+            this.setAuthorDisplayName(author, author.getDisplayName());
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: No need for the empty line above, since this whole chunk is setting up of the `RepoConfiguration`

File: src/main/java/reposense/model/Author.java [hunk 8]
```diff
@@ -8,21 +8,91 @@
 public class Author {
     public static final String UNKNOWN_AUTHOR_GIT_ID = "-";
 
+    private static final String MESSAGE_ILLEGAL_GIT_ID = "The provided Git ID, %s, contains illegal characters.";
+    private static final String MESSAGE_ILLEGAL_DISPLAY_NAME =
+            "The provided display name, %s, contains illegal characters.";
+    private static final String MESSAGE_ILLEGAL_AUTHOR_ALIAS =
+            "The provided author alias, %s,  contains illegal characters.";
+    private static final String MESSAGE_UNCOMMON_GLOB_PATTERN =
+            "The provided ignore glob, %s,  uses uncommon pattern.";
```
Comment (yamidark) [inline src/main/java/reposense/model/Author.java]: extra whitespace after comma for above 2 lines

File: src/main/java/reposense/model/RepoConfiguration.java [hunk 9]
```diff
@@ -126,24 +126,20 @@ public static void setFormatsToRepoConfigs(List<RepoConfiguration> configs, List
      * Clears authors information and use the information provided from {@code standaloneConfig}.
      */
     public void update(StandaloneConfig standaloneConfig) {
-        authorList.clear();
+        authorList = new ArrayList<>();
```
Comment (yamidark) [inline src/main/java/reposense/model/RepoConfiguration.java]: unnecessary change?

File: src/test/java/reposense/model/AuthorTest.java [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author(""));
+    }
+
+    @Test
+    public void constructor_illegalGitHubId_throwsIllegalArgumentException() {
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author("lithiumlkid & echo hi"));
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author("lithiumlkid ; echo hi"));
+        AssertUtil.assertThrows(IllegalArgumentException.class, () -> new Author("lithiumlkid | rm -rf /"));
+    }
+
```
Comment (yamidark) [inline src/test/java/reposense/model/AuthorTest.java]: add tests for the globs as well (using `setIgnoreGlobList`)?

File: src/main/java/reposense/model/RepoConfiguration.java (hunk 9 above)
Comment (eugenepeh) [reply src/main/java/reposense/model/RepoConfiguration.java]: this is a change necessary for me to use GitTestTemplate as it initializes authorList to be non-mutable by default.

File: src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json (hunk 3 above)
Comment (eugenepeh) [reply src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json]: in this file, only this line is invalid.
`"authorNames": ["Codee; ls"],`

File: src/test/java/reposense/model/StandaloneConfigTest.java (hunk 4 above)
Comment (eugenepeh) [reply src/test/java/reposense/model/StandaloneConfigTest.java]: that should belong to `StandaloneConfigJsonParserTest` and we can add that in other PR in future

File: src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json (hunk 3 above)
Comment (yamidark) [reply src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json]: Yea, but as they are other 'valid' `authorName`s here, it may be confusing for someone new to figure out which `authorName` for this file is invalid, hence I suggested should have just 1 author for each file.

Comment (eugenepeh) [reply src/test/resources/StandaloneConfigTest/invalidAuthorAliases_config.json]: Hmm, the thing is, it is not very accurate if I only have a single author. 
E.g. my regex can be checking for valid authors and not checking for invalid authors.
I will rename the file for it to be more intuitive.

File: src/test/java/reposense/model/AuthorTest.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        String[] ignoreGlobs = new String[] {"**.adoc", "collated/**"};
+        String[] moreIgnoreGlobs = new String[] {"**[!(.md)]", "C:\\Program Files\\**"};
+        List<String> ignoreGlobList = new ArrayList<>(Arrays.asList(ignoreGlobs));
+        ignoreGlobList.addAll(Arrays.asList(moreIgnoreGlobs));
+
+        author.setIgnoreGlobList(Arrays.asList(ignoreGlobs));
+        author.appendIgnoreGlobList(Arrays.asList(moreIgnoreGlobs));
+
+        Assert.assertEquals(4, author.getIgnoreGlobList().size());
+        Assert.assertTrue(author.getIgnoreGlobList().containsAll(ignoreGlobList));
```
[the remark this answers is not in the document]
Comment (eugenepeh) [reply src/test/java/reposense/model/AuthorTest.java]: ignoreGlobList already contains `moreIgnoreGlobs`

`ignoreGlobList.addAll(Arrays.asList(moreIgnoreGlobs));`

File: src/test/java/reposense/model/AuthorTest.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+    @Test(expected = IllegalArgumentException.class)
+    public void appendIgnoreGlobList_appendOrOperator_throwIllegalArgumentException() {
+        Author author = new Author("Tester");
+        String[] ignoreGlobs = new String[] {"**.adoc", "collated/**"};
+        String[] moreIgnoreGlobs = new String[] {"**[!(.md)] | rm -rf /", "C:\\Program Files\\**"};
+
+        author.setIgnoreGlobList(Arrays.asList(ignoreGlobs));
+        Assert.assertEquals(2, author.getIgnoreGlobList().size());
+        Assert.assertTrue(author.getIgnoreGlobList().containsAll(Arrays.asList(ignoreGlobs)));
```
[the remark this answers is not in the document]
Comment (eugenepeh) [reply src/test/java/reposense/model/AuthorTest.java]: hmm, currently there is no way to insert existing `ignoreGlobList` other than `setIgnoreGlobList` though.
This test also serves as a check for `UnsupportedOperationException`

File: src/test/java/reposense/model/AuthorTest.java [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+        author = new Author(gitId);
+        Assert.assertEquals(gitId, author.getGitId());
+    }
+
+    @Test
+    public void setIgnoreGlobList_validGlobRegex_success() {
+        Author author = new Author("Tester");
+        String[] ignoreGlobs = new String[] {"**.adoc", "collated/**"};
+
+        author.setIgnoreGlobList(Arrays.asList(ignoreGlobs));
```
[the remark this answers is not in the document]
Comment (eugenepeh) [reply src/test/java/reposense/model/AuthorTest.java]: oh yes, missed out this one, thanks!

File: src/test/java/reposense/model/AuthorTest.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (eugenepeh) [reply src/test/java/reposense/model/AuthorTest.java]: oh yes, I get what you mean. Make sense because `setIgnoreGlobList` can trigger `IllegalArgumentException` which makes the test pass anyway.

File: src/main/java/reposense/model/Author.java [hunk 14]
```diff
@@ -5,30 +5,103 @@
...
     public static final String UNKNOWN_AUTHOR_GIT_ID = "-";
 
-    private String gitId;
+    private static final String MESSAGE_ILLEGAL_GIT_ID = "The provided Git ID, %s, contains illegal characters.";
+    private static final String MESSAGE_ILLEGAL_DISPLAY_NAME =
+            "The provided display name, %s, contains illegal characters.";
+    private static final String MESSAGE_ILLEGAL_AUTHOR_ALIAS =
+            "The provided author alias, %s, contains illegal characters.";
+    private static final String MESSAGE_UNCOMMON_GLOB_PATTERN =
+            "The provided ignore glob, %s, uses uncommon pattern.";
```
[the remark this answers is not in the document]
Comment (eugenepeh) [reply src/main/java/reposense/model/Author.java]: Cause glob is a pattern rather than a string. Hence I named it pattern.
`uncommon` is because almost anything under the sun can be inserted as glob and it is still valid.
name is `illegal` because we dont accept `uncommon` as input, hence its an IllegalArgument to us.

File: src/main/java/reposense/model/Author.java [hunk 15]
```diff
@@ -5,30 +5,103 @@
...
+
+    private static final String COMMON_GLOB_REGEX = "^[-a-zA-Z0-9 _/\\\\*!{}\\[\\]!(),:.]*$";
+    private static final String NAME_VALIDATION_REGEX = "^[-a-zA-Z0-9 _/\\\\*]+$";
```
Comment (ongspxm) [inline src/main/java/reposense/model/Author.java]: shouldn't this 2 be dump inside the function itself? i mean it works, but...

Comment (eugenepeh) [reply src/main/java/reposense/model/Author.java]: For magic values, we usually represent them with a constant
- easier to maintain
- more readable with a meaningful name
